MCPClient: Fix package extraction

jrwdunham opened the pull request:
1. Sets packages' `currentlocation` value to the correct post-extraction (date-suffixed) location in the database.
2. Modifies `extract_contents.py::temporary_directory` to introduce a new delimiter (`'-AM-DATE-SFX-'`) between the package name and the date string suffix.
3. Modifies `has_packages.py::already_extracted` to strip the delimiter and date from the package path prior to searching for extracted files in the database that match that path in order to determine whether the package has already been extracted.

This was connected previously to archivematica/Issues#116, but now [PR 1275](https://github.com/artefactual/archivematica/pull/1275) should resolve that issue.


## Manual tests performed

I tested this by creating an AIP from a directory named `smallzipsanitizee/` which contained a .zip file named `small zip.zip` which, once extracted, contains `smallzip/small/small.txt`.

The resulting AIP has the following directory structure and crucially includes the renamed package file `small_zip.zip-AM-DATE-SFX-2018-08-28T16_53_09.602976_00_00`:

    tue5-5a94a630-8913-4b2b-96b6-aa3a918d6aba
    ├── bag-info.txt
    ├── bagit.txt
    ├── data
    │   ├── METS.5a94a630-8913-4b2b-96b6-aa3a918d6aba.xml
    │   ├── README.html
    │   ├── logs
    │   │   ├── fileFormatIdentification.log
    │   │   ├── filenameCleanup.log
    │   │   └── transfers
    │   │       └── tue5-f19352e5-d80d-4453-922c-1d709620a4fc
    │   │           └── logs
    │   │               ├── extractContents.log
    │   │               ├── fileFormatIdentification.log
    │   │               └── filenameCleanup.log
    │   └── objects
    │       ├── small_zip.zip
    │       │   └── smallzip
    │       │       └── small
    │       │           └── small.txt
    │       ├── small_zip.zip-AM-DATE-SFX-2018-08-28T16_53_09.602976_00_00
    │       └── submissionDocumentation
    │           └── transfer-tue5-f19352e5-d80d-4453-922c-1d709620a4fc
    │               └── METS.xml
    ├── manifest-sha256.txt
    └── tagmanifest-md5.txt

The METS file documents that package file in the physical structmap:

    <mets:structMap ID="structMap_1" LABEL="Archivematica default" TYPE="physical">
      <mets:div LABEL="tue5-5a94a630-8913-4b2b-96b6-aa3a918d6aba" TYPE="Directory" DMDID="dmdSec_1">
        <mets:div LABEL="objects" TYPE="Directory">
          <mets:div LABEL="small_zip.zip" TYPE="Directory">
            <mets:div LABEL="smallzip" TYPE="Directory">
              <mets:div LABEL="small" TYPE="Directory">
                <mets:div LABEL="small.txt" TYPE="Item">
                  <mets:fptr FILEID="file-2e823d72-b7ad-48bc-96a5-8a6bb26b67fd"/>
                </mets:div>
              </mets:div>
            </mets:div>
          </mets:div>
          <mets:div LABEL="small_zip.zip-AM-DATE-SFX-2018-08-28T16_53_09.602976_00_00" TYPE="Item">
            <mets:fptr FILEID="file-0d846675-bf86-46fb-a245-bb027bd4c4cc"/>
          </mets:div>
          <mets:div LABEL="submissionDocumentation" TYPE="Directory">
            <mets:div LABEL="transfer-tue5-f19352e5-d80d-4453-922c-1d709620a4fc" TYPE="Directory">
              <mets:div LABEL="METS.xml" TYPE="Item">
                <mets:fptr FILEID="file-c83d93b9-bdbb-4405-b2fb-a1a5ddc78d90"/>
              </mets:div>
            </mets:div>
          </mets:div>
        </mets:div>
      </mets:div>
    </mets:structMap>

And the corresponding `<mets:amdSec>` documents the original name correctly:

    <premis:originalName>%transferDirectory%objects/small zip.zip</premis:originalName>

jrwdunham committed:
MCPClient: Fix package extraction

1. Sets packages' ``currentlocation`` value to the correct post-extraction
   (date-suffixed) location in the database.
2. Modifies ``extract_contents.py::temporary_directory`` to introduces a new
   delimiter (``'-AM-DATE-SFX-'``) between the package name and the date string
   suffix.
3. Modifies ``has_packages.py::already_extracted`` to strip the delimiter and
   date from the package path prior to searching for extracted files in the
   database that match that path in order to determine whether the package has
   already been extracted.

Connected to archivematica/Issues#116

ross-spencer commented:
Hi @jrwdunham - Initial functional tests show that this might not capture all permutations. I have attached a zip that is failing on 'Characterize and Extract Metadata' with the following:
```
[Errno 2] No such file or directory: '/var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/2_again-df0ed1c8-9fd4-4f7c-938a-07aa0e53d94a/objects/one_file.zip-AM-DATE-SFX-2018-08-28T18:27:20.853989+00:00'Traceback (most recent call last):
  File "/src/MCPClient/lib/job.py", line 102, in JobContext
    yield
  File "/src/MCPClient/lib/clientScripts/store_file_modification_dates.py", line 65, in call
    main(transfer_uuid, shared_directory_path)
  File "/src/MCPClient/lib/clientScripts/store_file_modification_dates.py", line 52, in main
    transfer_file.modificationtime = get_modification_date(file_path)
  File "/src/MCPClient/lib/clientScripts/store_file_modification_dates.py", line 36, in get_modification_date
    mod_time = os.path.getmtime(file_path)
  File "/usr/lib/python2.7/genericpath.py", line 54, in getmtime
    return os.stat(filename).st_mtime
OSError: [Errno 2] No such file or directory: '/var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/2_again-df0ed1c8-9fd4-4f7c-938a-07aa0e53d94a/objects/one_file.zip-AM-DATE-SFX-2018-08-28T18:27:20.853989+00:00'
```
This file should be unzipped and placed into transfers to test (it will contain:
```
tree zip-zip
zip-zip
├── one_file_two.zip
├── one_file.txt
└── one_file.zip

0 directories, 3 files
```
[zip-zip.zip](https://github.com/artefactual/archivematica/files/2329405/zip-zip.zip)

ross-spencer commented on src/MCPClient/lib/clientScripts/extract_contents.py:
```diff
@@ -167,6 +167,8 @@ def main(job, transfer_uuid, sip_directory, date, task_uuid, delete=False):
         # the names we want to preserve in our PREMIS:originalLocation.
         temp_dir = temporary_directory(output_file_path, date)
         rename(output_file_path, temp_dir)
+        file_.currentlocation = temporary_directory(file_.currentlocation, date)
```
I wonder if we could extend `fileOperations.rename(...)` with an option to update the current location? It seems like the two could be linked in such a way, but I'm not sure where else in the code we currently perform something similar to this?

jrwdunham replied:
Yes, maybe this is a good idea. I have searched through the code and found the following:

- Places where `fileOperations.py::rename` is used:
  - extract_contents.py
    - `fileOperations.py::updateFileLocation` is NOW also used (post
      ce4e41b0f7811d0d96997703537878d13ee16076)
  - move_dspace_license_files_to_dspace_licenses.py
    - `fileOperations.py::updateFileLocation` IS also used
  - move_dspace_mets_files_to_dspace_mets.py
    - `fileOperations.py::updateFileLocation` IS also used
  - check_for_access_directory.py
    - `fileOperations.py::updateFileLocation` IS also used
  - move_transfer.py
    - `fileOperations.py::updateFileLocation` is NOT also used, BUT it is
      moving an entire transfer
  - move_sip.py
    - `fileOperations.py::updateFileLocation` is NOT also used, BUT it is
      moving an entire SIP

So if we were to call `fileOperations.py::updateFileLocation` from within `fileOperations.py::rename` then we would have to:

1. prevent the calling of `updateFileLocation` in the transfer/SIP cases; and
2. make sure that `updateFileLocation` is being used in the same way in the other four cases and add parameterization where it is not.

I'm not entirely sure it is worth the effort, but if you think it is I could do it.

This comment did, however, lead me to add a call to `updateFileLocation` in `extract_contents.py` which was needed: now the package files should have PREMIS events of type `'movement'`, which was definitely needed I think:

    <premis:eventType>movement</premis:eventType>
    <premis:eventDateTime>2018-08-28T22:39:00.360525+00:00</premis:eventDateTime>
    <premis:eventDetail/>
    <premis:eventOutcomeInformation>
      <premis:eventOutcome/>
      <premis:eventOutcomeDetail>
        <premis:eventOutcomeDetailNote>
          moved from="%transferDirectory%objects/one_file.zip"; moved to="%transferDirectory%objects/one_file.zip-AM-DATE-SFX-2018-08-28T22:39:00.168851+00:00"
        </premis:eventOutcomeDetailNote>
      </premis:eventOutcomeDetail>
    </premis:eventOutcomeInformation>

One thing I noticed was that the `'unpacking'` events of the unpacked files now reference the unmodified names of the packages that they were unpacked from. See below. I think this is ok, but let me know if you think otherwise.

    <premis:eventType>unpacking</premis:eventType>
    <premis:eventDateTime>2018-08-28T22:39:00.337993+00:00</premis:eventDateTime>
    <premis:eventDetail>
      Unpacked from: %transferDirectory%objects/one_file.zip (ac88a562-e26a-4b2b-b31b-02b27289515e)
    </premis:eventDetail>

ross-spencer commented on src/MCPClient/lib/clientScripts/has_packages.py:
```diff
@@ -24,14 +24,29 @@ def is_extractable(f):
         return False
 
 
+AM_DATE_DELIMITER = '-AM-DATE-SFX-'
```
Reading your rationale I like this approach. As to where values like this should sit, do you think there is potential to have them at a more central location across the codebase (archivematicaCommon?) or if there is a place where Archivematica 'special values'/keywords could exist so that they can be looked up? E.g. in validation operations to make sure user's don't accidentally transfer values with keywords in them - or that we know they're there and can work around them as appropriate? 

I don't want to expand the scope of this fix too much though, and I am still getting to grips with the impact of what might happen say if we wanted to actually transfer a file with a `processingMCP.xml` or `metadata` folder in them, but have those treated as objects not special files/dirs in the system. 

This might just be gold plating at this point and have no practical use, so please feel free to ignore.

jrwdunham replied:
I see what you mean, it would be good for future work to define these in a central location. However, I don't see anything in archivematicaCommon/lib/ or MCPClient/lib/clientScripts/ where a precedent for something like this has been set. I guess we could introduce a constants.py file somewhere, but I'm inclined to leave this as is for now.

jrwdunham replied:
I see what's bothering you though. The interconnectedness of `already_extracted` and `suffix_date` is something that should be enforced or at least more explicitly expressed than in shared use of a constant. I'm wondering if this might be a case where property-based testing (Hypothesis: https://hypothesis.readthedocs.io/en/latest/) might be useful...

ross-spencer replied:
I like the look of Hypothesis! yeah, I guess it's a combination of the both your points here. I suspect for testing and for users we could benefit from something like `constants.py` as well as an update to the docs, just explaining that various names might cause issues, but also, linking out to the sections of the docs that describe how to take advantage of them. Definitely out of scope for this issue!

ross-spencer reviewed (changes requested):
Two comments @jrwdunham which i'm interested in your thoughts on for this PR. Otherwise, as noted, once the other potential change is addressed from testing i like this approach and think this looks good.

jrwdunham committed:
MCPClient: add extracted package movement event

Causes a 'movement' event to be added to the database (and ultimately
the METS) after extracting a package because now (post PR 1178)
extracting a package means moving it on disk (i.e., changing its name).

jrwdunham commented:
@ross-spencer I tried creating an AIP from your `zip-zip.zip` (after unpacking it) and it seems to have worked correctly for me. No problems under `Characterize and extract metadata` or elsewhere. See the resulting structmap below.

    <mets:structMap ID="structMap_1" LABEL="Archivematica default" TYPE="physical">
      <mets:div LABEL="tue4-f51fabc5-c387-43ef-94b5-a40afda350a1" TYPE="Directory" DMDID="dmdSec_1">
        <mets:div LABEL="objects" TYPE="Directory">
          <mets:div LABEL="one_file.txt" TYPE="Item">
            <mets:fptr FILEID="file-a14e646a-2ee5-45da-abec-6d5cdfed0c76"/>
          </mets:div>
          <mets:div LABEL="one_file.zip" TYPE="Directory">
            <mets:div LABEL="one_file" TYPE="Directory">
              <mets:div LABEL="one_file.txt" TYPE="Item">
                <mets:fptr FILEID="file-94154b16-c325-4f41-9b01-872732bf5646"/>
              </mets:div>
            </mets:div>
          </mets:div>
          <mets:div LABEL="one_file.zip-AM-DATE-SFX-2018-08-28T22_39_00.168851_00_00" TYPE="Item">
            <mets:fptr FILEID="file-ac88a562-e26a-4b2b-b31b-02b27289515e"/>
          </mets:div>
          <mets:div LABEL="one_file_two.zip" TYPE="Directory">
            <mets:div LABEL="one_file.txt" TYPE="Item">
              <mets:fptr FILEID="file-14fb128a-d44f-4dc2-94b3-db951a6dbb60"/>
            </mets:div>
            <mets:div LABEL="one_file.zip" TYPE="Directory">
              <mets:div LABEL="one_file" TYPE="Directory">
                <mets:div LABEL="one_file.txt" TYPE="Item">
                  <mets:fptr FILEID="file-64105e5e-abaf-4183-bd8d-7a72bbe805bf"/>
                </mets:div>
              </mets:div>
            </mets:div>
            <mets:div LABEL="one_file.zip-AM-DATE-SFX-2018-08-28T22_39_05.625402_00_00" TYPE="Item">
              <mets:fptr FILEID="file-f134feb5-00a3-4ecc-8832-2e121f27c3f2"/>
            </mets:div>
          </mets:div>
          <mets:div LABEL="one_file_two.zip-AM-DATE-SFX-2018-08-28T22_39_00.168851_00_00" TYPE="Item">
            <mets:fptr FILEID="file-385b3dee-e24a-4150-bb7a-b17dc1dea841"/>
          </mets:div>
          <mets:div LABEL="submissionDocumentation" TYPE="Directory">
            <mets:div LABEL="transfer-tue4-2510b75a-59da-46f6-bc57-0f4a72e15456" TYPE="Directory">
              <mets:div LABEL="METS.xml" TYPE="Item">
                <mets:fptr FILEID="file-467285fc-27a4-4fca-bab5-dc0cb8ac8266"/>
              </mets:div>
            </mets:div>
          </mets:div>
        </mets:div>
      </mets:div>
    </mets:structMap>

Is it possible that you did not run `docker-compose up -d --force-recreate --no-deps archivematica-mcp-client` before testing?

jrwdunham commented:
@ross-spencer maybe what you saw is a variable bug where it arises depending on the order of iteration in a for-loop. I'm thinking that maybe the string matching logic in `already_extracted ` is giving false positives because some of your zip file paths are prefixes of others. Something to check...

ross-spencer commented:
Hi @jrwdunham I was worried I hadn't rebuilt the image properly, but my normal process means I usually run: `docker-compose up -d --force-recreate --build`.

Can i ask what happens if you try that downloaded zip I provided, twice in a row. 

I am seeing:

* First will pass
* Second will fail

With differences in Sanitize Names:

**First:**
```
Sanitized name: %transferDirectory%objects/one_file_two.zip-AM-DATE-SFX-2018-08-29T15:26:48.878378+00:00  ->  %transferDirectory%objects/one_file_two.zip-AM-DATE-SFX-2018-08-29T15_26_48.878378_00_00
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/objects/one_file_two.zip/one_file.zip
Sanitized name: %transferDirectory%objects/one_file.zip-AM-DATE-SFX-2018-08-29T15:26:48.878378+00:00  ->  %transferDirectory%objects/one_file.zip-AM-DATE-SFX-2018-08-29T15_26_48.878378_00_00
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/objects/one_file.txt
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/objects/one_file.zip/one_file/one_file.txt
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/objects/one_file_two.zip/one_file.txt
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/objects/one_file.zip/one_file/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/logs/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/metadata/submissionDocumentation/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/metadata/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/logs/fileMeta/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/objects/one_file_two.zip/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh1-b8a3e6e9-b90d-46c9-be04-34fffc3c99f2/objects/one_file.zip/
```
**Second:**
```
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file.zip/one_file/one_file.txt
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file.txt
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file_two.zip-AM-DATE-SFX-2018-08-29T15:26:48.878378+00:00
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file_two.zip/one_file.txt
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file.zip-AM-DATE-SFX-2018-08-29T15:26:48.878378+00:00
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file_two.zip/one_file.zip
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/logs/fileMeta/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/logs/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file.zip/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/metadata/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file_two.zip/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file.zip/one_file/
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/metadata/submissionDocumentation/
```

If I sort those the only differences are:

**First:**
```
Sanitized name: %transferDirectory%objects/one_file_two.zip-AM-DATE-SFX-2018-08-29T15:26:48.878378+00:00  ->  %transferDirectory%objects/one_file_two.zip-AM-DATE-SFX-2018-08-29T15_26_48.878378_00_00

Sanitized name: %transferDirectory%objects/one_file.zip-AM-DATE-SFX-2018-08-29T15:26:48.878378+00:00  ->  %transferDirectory%objects/one_file.zip-AM-DATE-SFX-2018-08-29T15_26_48.878378_00_00
```
**Second:**
```
No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file_two.zip-AM-DATE-SFX-2018-08-29T15:26:48.878378+00:00

No sanitization found for /var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zzgh2-3f16f48c-004b-49a4-8cb6-d78069d195e6/objects/one_file.zip-AM-DATE-SFX-2018-08-29T15:26:48.878378+00:00
```

I haven't been able to understand why this pattern yet but the difference in sanitize names seems to affet what happens further up the chain:

![image](https://user-images.githubusercontent.com/1880412/44798952-6c86d100-ab78-11e8-909b-5f603d462bda.png)

This is what I see when I compare the error log with the failed transfers folder in a previous attempt:
```
root@artefactual:/tmp/am-pipeline-data/failed/zz2-1add1f15-2ac9-4fcb-9ccf-1974b629746a# tree
.
├── logs
│   ├── extractContents.log
│   ├── fileFormatIdentification.log
│   ├── fileMeta
│   └── filenameCleanup.log
├── metadata
│   ├── directory_tree.txt
│   └── submissionDocumentation
│       └── METS.xml
├── objects
│   ├── one_file_two.zip
│   │   ├── one_file.txt
│   │   ├── one_file.zip
│   │   │   └── one_file
│   │   │       └── one_file.txt
│   │   └── one_file.zip-AM-DATE-SFX-2018-08-29T14_58_55.097044_00_00
│   └── one_file_two.zip-AM-DATE-SFX-2018-08-29T14_58_50.648566_00_00
└── processingMCP.xml

8 directories, 10 files
```

```
[Errno 2] No such file or directory: '/var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zz2-1add1f15-2ac9-4fcb-9ccf-1974b629746a/objects/one_file_two.zip/one_file.zip-AM-DATE-SFX-2018-08-29T14:57:44.685339+00:00'Traceback (most recent call last):
  File "/src/MCPClient/lib/job.py", line 102, in JobContext
    yield
  File "/src/MCPClient/lib/clientScripts/store_file_modification_dates.py", line 65, in call
    main(transfer_uuid, shared_directory_path)
  File "/src/MCPClient/lib/clientScripts/store_file_modification_dates.py", line 52, in main
    transfer_file.modificationtime = get_modification_date(file_path)
  File "/src/MCPClient/lib/clientScripts/store_file_modification_dates.py", line 36, in get_modification_date
    mod_time = os.path.getmtime(file_path)
  File "/usr/lib/python2.7/genericpath.py", line 54, in getmtime
    return os.stat(filename).st_mtime
OSError: [Errno 2] No such file or directory: '/var/archivematica/sharedDirectory/watchedDirectories/workFlowDecisions/extractPackagesChoice/zz2-1add1f15-2ac9-4fcb-9ccf-1974b629746a/objects/one_file_two.zip/one_file.zip-AM-DATE-SFX-2018-08-29T14:57:44.685339+00:00'
```
![image](https://user-images.githubusercontent.com/1880412/44799201-fc2c7f80-ab78-11e8-84cf-e36b0b8d71f3.png)

jrwdunham committed:
MCPClient: fix package date suffixing and matching

- `suffix_date` no longer uses the global (mutable) dict
  `file_path_cache` as a cache. When it did so it could return bad
  cached values (set during previous transfers) for identical paths.
- `already_extracted` now adds a trailing forward slash to the package
  name when it searches for files extracted from that package. Without
  doing this it may be possible to get false positive matches when
  another package in the same directory contains as a prefix the name of the
  first package.

jrwdunham commented:
@ross-spencer thanks for all the details and for the test transfer which has revealed issues in this PR!

I think I've figured out the fundamental problem: the `suffix_date` function was using a global (mutable) dict `file_path_cache` as a cache. The problem with this is that it was then returning incorrect cached values for the date-suffixed version of a package path. Those cached values were set during previous transfers. This only became a possible issue when we started importing client scripts as Python modules instead of forking new processes for them.

@sevein pointed to this commit as related: https://github.com/artefactual/archivematica/commit/afd976222c60269d62c51e4147f7831b5de46a33. He also found this as an example of how we might find cases in our code where functions are accessing and modifying global mutable values (so that we can pro-actively refactor this problem away): https://stackoverflow.com/questions/33160744/detect-all-global-variables-within-a-python-function/33160791#33160791

At any rate, can you test this with your challenging transfer again now that the third commit has been added? I was able to make that transfer work three times in a row, so I think it's fixed now.

ross-spencer commented on src/MCPClient/lib/clientScripts/extract_contents.py:
```diff
@@ -28,13 +33,8 @@
-        path = file_path + '-' + date
-        file_path_cache[file_path] = path
-        return path
+def suffix_date(file_path, date):
```
The function is pretty self-explanatory, but do you think we should add a docstring as well?

ross-spencer reviewed (approved):
Hi @joel-simpson this looks good. I just left a single comment asking about a function docstring. Before merging, I think this is one that we might just want @sromkey's opinion as product owner. I think mainly just to manage expectations on what an AIP will eventually look like when packages aren't deleted (namely, with the `-AM-DATE-SFX` shown below.) And I think the issue remains open with acceptance criteria for the docs to be updated as well, but that's not necessarily something you're responsible for. (I think anyway! There was a little on this in the weekly while you were away so just trying to remember as accurately as possible as the process moves a bit)  
```
ross-spencer@artefactual:~/Desktop/Artefactual/AIPS/1.7.1/qa-1.x$ tree onezip-93b3308e-5bd4-4b5a-b9d0-69874259fcb1
onezip-93b3308e-5bd4-4b5a-b9d0-69874259fcb1
├── bag-info.txt
├── bagit.txt
├── data
│   ├── logs
│   │   ├── fileFormatIdentification.log
│   │   ├── filenameCleanup.log
│   │   └── transfers
│   │       └── onezip-5dcd2dab-6cad-410e-a627-486d95d54329
│   │           └── logs
│   │               ├── extractContents.log
│   │               ├── fileFormatIdentification.log
│   │               └── filenameCleanup.log
│   ├── METS.93b3308e-5bd4-4b5a-b9d0-69874259fcb1.xml
│   ├── objects
│   │   ├── metadata
│   │   │   └── transfers
│   │   │       └── onezip-5dcd2dab-6cad-410e-a627-486d95d54329
│   │   │           └── directory_tree.txt
│   │   ├── one_file.zip
│   │   │   └── one_file
│   │   │       └── one_file.txt
│   │   ├── one_file.zip-AM-DATE-SFX-2018-08-30T02_33_17.349607_00_00
│   │   └── submissionDocumentation
│   │       └── transfer-onezip-5dcd2dab-6cad-410e-a627-486d95d54329
│   │           └── METS.xml
│   └── README.html
├── manifest-sha256.txt
└── tagmanifest-md5.txt

13 directories, 15 files
```

ross-spencer commented:
RE: https://github.com/artefactual/archivematica/pull/1245#issuecomment-417092171 I have logged #130 so that we can discuss it in the backlog grooming meeting later today.

jrwdunham commented:
@ross-spencer I think your intention was to reference this issue https://github.com/archivematica/Issues/issues/130

jrwdunham commented:
I am closing this PR on account of @sromkey's comment at https://github.com/archivematica/Issues/issues/116#issuecomment-417401572. I will open a new PR that fixes the issue by undoing the package renaming that was introduced by https://github.com/artefactual/archivematica/pull/1178.
